### PR TITLE
Dendrites as Bias and Gates

### PR DESCRIPTION
**Pull request by mvacaporale**
This introduces a `DendriticSegments` module which combines `DendriteSegments` and `SparseWeights` modules. Dendrites are applied as a bias and a subclass `GatingDendriticSegments` applies them as a gating mechanism.

A new subfolder `frameworks.dendrites` has been introduced and some basic unit tests have been included.

Please let me know if there are any further unit tests that should be included.

A future PR will implement something similiar for 2d (conv) weights.

### CODE REVIEW

File: nupic/research/frameworks/dendrites/dendrite_segments.py [hunk 1]
```diff
@@ -60,20 +60,20 @@ def __init__(self, num_units, num_segments, dim_context, sparsity, bias=None):
         self.sparsity = sparsity
 
         # TODO: Use named dimensions.
```
Comment (lucasosouza) [inline nupic/research/frameworks/dendrites/dendrite_segments.py:62]: You mentioned in another PR you didn't want to use named dimensions. If that is the case, then better to remove this TODO

File: nupic/research/frameworks/dendrites/dendritic_weights.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    def forward(self, x, context):
+        """
+        Compute of linear layer and apply output of dendrite segments.
+        """
+        y = super().forward(x)
+        dendrite_activations = self.segments(context)  # num_units x num_segments
+        return self.apply_dendrites(y, dendrite_activations)
+
+
+class GatingDendriticWeights(DendriticWeights):
```
Comment (lucasosouza) [inline nupic/research/frameworks/dendrites/dendritic_weights.py:82]: Is it worth having a separate class just for this?

**Review by lucasosouza — APPROVED** [review]

File: nupic/research/frameworks/dendrites/dendritic_weights.py (hunk 2 above)
Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/dendritic_weights.py:82]: It is I think. We may eventually canonicalize the way we apply dendrites, but I wanted to design the modules in a way to experiment with many formats. For instance, we may want to apply a sigmoid prior to gating, or maybe we'd want to add a bias term after applying the max. Ideally, any variants should be quick to try. This gating class is just an example of how one may go about doing that.

Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/dendritic_weights.py:82]: Note, my goal here is to make these classes easy to iterate off of.

Comment (karangrewal) [inline nupic/research/frameworks/dendrites/dendritic_weights.py:82]: If I understand correctly,

`DendriticWeights` --> applies dendrite activations as bias,
`GatingDendriticWeights` --> applies dendrite activations as multiplicative value,

so why is one a subclass of the other?

Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/dendritic_weights.py:82]: That's correct.

See this comment above:
https://github.com/numenta/nupic.research/pull/356#discussion_r497548307

I don't expect these classes to be the final way we organize the code, but a simple template to iterate from. The `apply_dendrites` function was something, in particular, I wanted to be easily changed.

**Review by karangrewal — APPROVED** [review]

File: nupic/research/frameworks/dendrites/dendritic_weights.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    def apply_dendrites(self, y, dendrite_activations):
+        """Apply dendrites as a gating mechanism."""
+        return y * dendrite_activations.max(dim=2).values  # max along each segment
```
Comment (subutai) [inline nupic/research/frameworks/dendrites/dendritic_weights.py:85]: Shouldn't there be a sigmoid here?

File: tests/unit/frameworks/dendrites/dendrites_test.py [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+        num_off = (dendrite_weights == 0).sum().item()
+        actual_dendrite_sparsity = num_off / dendrite_weights.numel()
+        self.assertEqual(dendrite_sparsity, actual_dendrite_sparsity)
+
+        num_off = (linear_weights == 0).sum().item()
+        actual_linear_sparsity = num_off / linear_weights.numel()
+        self.assertEqual(linear_sparsity, actual_linear_sparsity)
+
+    def test_gating_forward(self):
+        """Validate shape of forward output."""
```
Comment (subutai) [inline tests/unit/frameworks/dendrites/dendrites_test.py:147]: We should add a test where the gating units actually output the expected values.

File: tests/unit/frameworks/dendrites/dendrites_test.py [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+from nupic.research.frameworks.dendrites import (
+    DendriteSegments,
+    DendriticWeights,
+    GatingDendriticWeights,
+)
+
+
+class DendriteSegmentsTests(unittest.TestCase):
+
+    def test_forward(self):
```
Comment (subutai) [inline tests/unit/frameworks/dendrites/dendrites_test.py:35]: We should add a test where the dendrites-as-bias units actually output the expected values. Please see https://github.com/numenta/nupic.torch/blob/master/tests/k_winners_test.py as an example.

**Review by subutai — CHANGES_REQUESTED** [review]

File: nupic/research/frameworks/dendrites/dendritic_weights.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+        self.rezero_weights()
+
+    def rezero_weights(self):
+        super().rezero_weights()
+        if self.segments is not None:  # only none at beggining of init
+            self.segments.rezero_weights()
+
+    def apply_dendrites(self, y, dendrite_activations):
+        """Apply dendrites as a bias."""
+        return y + dendrite_activations.max(dim=2).values  # max along each segment
```
[the remark this answers is not in the document]
Comment (mvacaporale) [reply nupic/research/frameworks/dendrites/dendritic_weights.py:71]: I've check this with a simple test
```
t = torch.tensor([1., 2., 3.], requires_grad=True)
y = torch.max(t, dim=0).values
loss = y.sum()
loss.backward()
print(t.grad)
>>> tensor([0., 0., 1.])
```

The gradient does work the way we need it. There's a small discussion about that here:
https://discuss.pytorch.org/t/confused-about-torch-max-and-gradient/14283

Essentially, the max operator keeps the elements through an identity operation, so the gradients are preserved.

**Comment by mvacaporale** [comment]
@subutai How's `BiasingDendriticLayer` and `GatingDendriticLayer` for naming?

**Comment by mvacaporale** [comment]
@subutai I've addressed all your comments and buffed up the tests. Please see the latest commits.

**Comment by subutai** [comment]
> @subutai How's `BiasingDendriticLayer` and `GatingDendriticLayer` for naming?

Sounds good - let's go with that for now!

**Review by subutai — APPROVED** [review]
Great, thanks. Looks good. Please merge - I'll need to go through it in detail later.
